crates/core_arch/src/mips/msa/macros.rs

```diff
@@ -1,5 +1,35 @@
 //! Utility macros.
 
+macro_rules! static_assert_imm_s5 {
+    ($imm:ident) => {
+        let _ = $crate::core_arch::macros::ValidateConstImm::<$imm, -16, 16>::VALID;
+    };
+}
+
+macro_rules! static_assert_imm_s10 {
+    ($imm:ident) => {
+        let _ = $crate::core_arch::macros::ValidateConstImm::<$imm, -512, 512>::VALID;
+    };
+}
+
+macro_rules! static_assert_imm_s11 {
+    ($imm:ident) => {
+        let _ = $crate::core_arch::macros::ValidateConstImm::<$imm, -1024, 1023>::VALID;
+    };
+}
+
+macro_rules! static_assert_imm_s12 {
+    ($imm:ident) => {
+        let _ = $crate::core_arch::macros::ValidateConstImm::<$imm, -2048, 2045>::VALID;
+    };
+}
+
+macro_rules! static_assert_imm_s13 {
```

Comment:
For the intrinsics using these there are 2 conditions:
- Range is `-4096` to `4095`.
- Immediate is a multiple of 8.

It may be better to split these 2 conditions into separate `static_assert`s.

Reply:
This macro `*_s{5,11,12,13}` are unused in this PR, but they will be used in patch 2 and 3.
I will add `static_assert_multiply_of` macro for asserting multiple issue.

Reply:
Just use a normal `static_assert!` for that. The specific `static_assert_imm*` macros are mainly used to avoid duplication.

Reply:
Moved *imm_s to subsequent patches.